### PR TITLE
feat(ONNXToTosa): legalizations for PRelu, Selu, Softplus, ThresholdedRelu.

### PR DESCRIPTION
**Pull request by ttjost**

### CODE REVIEW

**Review by mgehre-amd — CHANGES_REQUESTED** [review]
Please fix the lowering for thresholded relu

File: test/mlir/conversion/onnx_to_tosa/Math/Elementwise.mlir [hunk 1]
```diff
@@ -166,6 +166,96 @@ func.func @test_leaky_relu(%arg0: tensor<13x21x3xf32>) -> tensor<13x21x3xf32> {
...
+// CHECK:           [[VAR_3_:%.+]] = "tosa.log"([[VAR_2_]]) : (tensor<13x21x3xf32>) -> tensor<13x21x3xf32>
+}
+
+// -----
+
+func.func @test_thresholdedrelu(%arg0: tensor<13x21x3xf32>) -> tensor<13x21x3xf32> {
+  %0 = "onnx.ThresholdedRelu"(%arg0) {alpha = 0.5  : f32} : (tensor<13x21x3xf32>) -> tensor<13x21x3xf32>
+  func.return %0 : tensor<13x21x3xf32>
+// CHECK-LABEL: test_thresholdedrelu
+// CHECK:           "tosa.clamp"(%arg0) <{max_fp = 3.40282347E+38 : f32, max_int = 2147483647 : i64, min_fp = 5.000000e-01 : f32, min_int = 0 : i64}> : (tensor<13x21x3xf32>) -> tensor<13x21x3xf32>
```
[the remark this answers is not in the document]
Comment (ttjost) [reply test/mlir/conversion/onnx_to_tosa/Math/Elementwise.mlir:245]: oh, you're right.

**Review by mgehre-amd — APPROVED** [review]

File: src/Conversion/ONNXToTOSA/Math/Elementwise.cpp [hunk 2]
```diff
@@ -241,16 +241,16 @@ class ONNXReluOpLoweringToTOSA : public OpConversionPattern<ONNXReluOp> {
...
   auto loc = op->getLoc();
   TosaBuilder tosaBuilder(rewriter, loc);
-  Value constZero = tosaBuilder.getSplattedConst(0.0, outputType.getShape());
+  Value constZero = tosaBuilder.getSplattedConst(
+      0.0, outputType.getShape(), outputType.getElementType());
 
   auto mul = tosa::CreateOpAndInfer<mlir::tosa::MulOp>(rewriter, op->getLoc(),
-      outputType, input,
-      tosaBuilder.getSplattedConst(alpha, outputType.getShape()),
+      outputType, input, alphaOrSlope,
```
Comment (TinaAMD) [inline src/Conversion/ONNXToTOSA/Math/Elementwise.cpp:253]: Does this automatically handle (rank-)broadcasting? Can you add a test with a PRelu where the slope requires broadcasting?

File: src/Conversion/ONNXToTOSA/Math/Elementwise.cpp [hunk 3]
```diff
@@ -272,9 +272,9 @@ class ONNXLeakyReluOpLoweringToTOSA
       ConversionPatternRewriter &rewriter) const override {
 
     auto outputType = op.getResult().getType().cast<TensorType>();
```
Comment (TinaAMD) [inline src/Conversion/ONNXToTOSA/Math/Elementwise.cpp:274]: Maybe also use the typeconverter on the elementtype (and you can re-use the result below in the `getSplattedConst` call.

File: src/Conversion/ONNXToTOSA/Math/Elementwise.cpp (hunk 2 above)
Comment (TinaAMD) [reply src/Conversion/ONNXToTOSA/Math/Elementwise.cpp:253]: Maybe create the mul with `tosaBuilder.binaryOp`?

File: src/Conversion/ONNXToTOSA/Math/Elementwise.cpp [hunk 4]
```diff
@@ -468,6 +472,118 @@ class ONNXHardSigmoidOpLoweringToTOSA
...
+        tosaBuilder.getSplattedConst(adaptor.getAlpha().convertToDouble(),
+            outputType.getShape(), outputType.getElementType());
+    Value gamma =
+        tosaBuilder.getSplattedConst(adaptor.getGamma().convertToDouble(),
+            outputType.getShape(), outputType.getElementType());
+    Value constZero = tosaBuilder.getSplattedConst(
+        0.0, outputType.getShape(), outputType.getElementType());
+
+    Value exp = tosaBuilder.unaryOp<mlir::tosa::ExpOp>(input);
+    Value expTimesAlpha = tosaBuilder.mul(exp, alpha);
```
Comment (TinaAMD) [inline src/Conversion/ONNXToTOSA/Math/Elementwise.cpp:544]: When do we use `tosaBuilder.mul` and when `tosaBuilder.binaryOp<mlir::tosa::MulOp>`?

File: src/Conversion/ONNXToTOSA/Math/Elementwise.cpp [hunk 5]
```diff
@@ -468,6 +472,118 @@ class ONNXHardSigmoidOpLoweringToTOSA
...
+
+    TosaBuilder tosaBuilder(rewriter, op->getLoc());
+    auto alpha =
+        tosaBuilder.getSplattedConst(adaptor.getAlpha().convertToDouble(),
+            outputType.getShape(), outputType.getElementType());
+    auto zero = tosaBuilder.getSplattedConst(
+        0.0, outputType.getShape(), outputType.getElementType());
+
+    auto greater = tosaBuilder.greater(input, alpha);
+    tosa::CreateReplaceOpAndInfer<mlir::tosa::SelectOp>(
```
Comment (TinaAMD) [inline src/Conversion/ONNXToTOSA/Math/Elementwise.cpp:581]: Similar thing for `CreateReplaceOpAndInfer`: When do we prefer this over building the op + `replaceOp`?

File: test/mlir/conversion/onnx_to_tosa/Math/Elementwise.mlir [hunk 6]
```diff
@@ -157,13 +157,119 @@ func.func @test_leaky_relu(%arg0: tensor<13x21x3xf32>) -> tensor<13x21x3xf32> {
...
+}
+
+// -----
+
+func.func @test_prelu(%arg0: tensor<13x21x3xf32>, %arg1: tensor<13x21x3xf32>) -> tensor<13x21x3xf32> {
+  %0 = "onnx.PRelu"(%arg0, %arg1) : (tensor<13x21x3xf32>, tensor<13x21x3xf32>) -> tensor<13x21x3xf32>
+  func.return %0 : tensor<13x21x3xf32>
+// CHECK-LABEL: test_prelu
+// CHECK-DAG:       [[VAR_0_:%.+]] = "tosa.const"() <{value = dense<0.000000e+00> : tensor<13x21x3xf32>}>
+// CHECK-DAG:       [[VAR_1_:%.+]] = "tosa.mul"(%arg0, [[PARAM_1_]]) <{shift = 0 : i32}>
```
Comment (TinaAMD) [inline test/mlir/conversion/onnx_to_tosa/Math/Elementwise.mlir:185]: `[[PARAM_1_]]` seems to be undefined.

File: test/mlir/conversion/onnx_to_tosa/Math/Elementwise.mlir [hunk 7]
```diff
@@ -157,13 +157,119 @@ func.func @test_leaky_relu(%arg0: tensor<13x21x3xf32>) -> tensor<13x21x3xf32> {
...
+// CHECK:           [[VAR_2_:%.+]] = "tosa.greater_equal"(%arg0, [[VAR_0_]])
+// CHECK:           [[VAR_3_:%.+]] = "tosa.select"([[VAR_2_]], %arg0, [[VAR_1_]])
+}
+
+func.func @test_prelu_bf16(%arg0: tensor<13x21x3xbf16>, %arg1: tensor<13x21x3xbf16>) -> tensor<13x21x3xbf16> {
+  %0 = "onnx.PRelu"(%arg0, %arg1) : (tensor<13x21x3xbf16>, tensor<13x21x3xbf16>) -> tensor<13x21x3xbf16>
+  func.return %0 : tensor<13x21x3xbf16>
+// CHECK-LABEL: test_prelu_bf16
+// CHECK-DAG:       [[VAR_0_:%.+]] = "tosa.const"() <{value = dense<0.000000e+00> : tensor<13x21x3xbf16>}> : () -> tensor<13x21x3xbf16>
+// CHECK-DAG:       [[VAR_1_:%.+]] = "tosa.mul"(%arg0, [[PARAM_1_]]) <{shift = 0 : i32}>
```
Comment (TinaAMD) [inline test/mlir/conversion/onnx_to_tosa/Math/Elementwise.mlir:195]: [[PARAM_1_]] seems to be undefined.

File: test/mlir/conversion/onnx_to_tosa/Math/Elementwise.mlir [hunk 8]
```diff
@@ -157,13 +157,119 @@ func.func @test_leaky_relu(%arg0: tensor<13x21x3xf32>) -> tensor<13x21x3xf32> {
...
+// CHECK-DAG:       [[VAR_1_:%.+]] = "tosa.const"() <{value = dense<1.000000e+00> : tensor<13x21x3xf32>}>
+// CHECK:           [[VAR_2_:%.+]] = "tosa.add"([[VAR_0_]], [[VAR_1_]])
+// CHECK:           [[VAR_3_:%.+]] = "tosa.log"([[VAR_2_]])
+// CHECK:           return [[VAR_3_]] : tensor<13x21x3xf32>
+}
+
+// -----
+
+func.func @test_thresholdedrelu(%arg0: tensor<13x21x3xf32>) -> tensor<13x21x3xf32> {
+  %0 = "onnx.ThresholdedRelu"(%arg0) {alpha = 0.5  : f32} : (tensor<13x21x3xf32>) -> tensor<13x21x3xf32>
```
Comment (TinaAMD) [inline test/mlir/conversion/onnx_to_tosa/Math/Elementwise.mlir:250]:
```suggestion
  %0 = "onnx.ThresholdedRelu"(%arg0) {alpha = 0.5 : f32} : (tensor<13x21x3xf32>) -> tensor<13x21x3xf32>
```

File: test/mlir/conversion/onnx_to_tosa/Math/Elementwise.mlir [hunk 9]
```diff
@@ -157,13 +157,119 @@ func.func @test_leaky_relu(%arg0: tensor<13x21x3xf32>) -> tensor<13x21x3xf32> {
...
+  func.return %0 : tensor<13x21x3xf32>
+// CHECK-LABEL: test_thresholdedrelu
+// CHECK-DAG:       [[VAR_0_:%.+]] = "tosa.const"() <{value = dense<5.000000e-01>
+// CHECK-DAG:       [[VAR_1_:%.+]] = "tosa.const"() <{value = dense<0.000000e+00>
+// CHECK:           [[VAR_2_:%.+]] = "tosa.greater_equal"(%arg0, [[VAR_0_]])
```
Comment (TinaAMD) [inline test/mlir/conversion/onnx_to_tosa/Math/Elementwise.mlir:255]: Doesn't this need to be `greater`? Otherwise we would get the wrong result for `0.5`, right?

**Review by TinaAMD — CHANGES_REQUESTED** [review]
Very nice-to-read lowerings!

File: src/Conversion/ONNXToTOSA/Math/Elementwise.cpp (hunk 4 above)
Comment (ttjost) [reply src/Conversion/ONNXToTOSA/Math/Elementwise.cpp:544]: we can't `binaryOp<MulOp>`, because may need a `shift` attribute.

File: test/mlir/conversion/onnx_to_tosa/Math/Elementwise.mlir (hunk 9 above)
Comment (ttjost) [reply test/mlir/conversion/onnx_to_tosa/Math/Elementwise.mlir:255]: Yes, I am actually using `greater`. Really weird that it is coming out as `greater_equal`.

File: src/Conversion/ONNXToTOSA/DialectBuilder.cpp [hunk 10]
```diff
@@ -272,6 +254,22 @@ template Value TosaBuilder::binaryOp<mlir::tosa::SubOp>(
 template Value TosaBuilder::binaryOp<mlir::tosa::PowOp>(
     mlir::Value &lhs, mlir::Value &rhs);
 
+template <typename T>
+Value TosaBuilder::unaryOp(mlir::Value &input) {
+  auto inputType = input.getType().cast<ShapedType>();
+  Type newValueType = RankedTensorType::get(
+      llvm::SmallVector<int64_t, 4>(inputType.getRank(), ShapedType::kDynamic),
+      inputType.getElementType());
```
[the remark this answers is not in the document]
Comment (ttjost) [reply src/Conversion/ONNXToTOSA/DialectBuilder.cpp:262]: I think we are checking the types when we are doing the conversion, so in theory, we don't need to check it again, since ONNX and tosa types are equivalent.

File: test/mlir/conversion/onnx_to_tosa/Math/Elementwise.mlir (hunk 9 above)
Comment (TinaAMD) [reply test/mlir/conversion/onnx_to_tosa/Math/Elementwise.mlir:255]: It is the same for Selu, but there (I think) it doesn't make a difference because for the `x = 0` corner case we get `0` for both calculations.

File: src/Conversion/ONNXToTOSA/DialectBuilder.cpp (hunk 10 above)
[the remark this answers is not in the document]
Comment (ttjost) [reply src/Conversion/ONNXToTOSA/DialectBuilder.cpp:262]: Also, I don't think we can use the typeconverter because it is only available in the `ConversionPattern` class and it would restrict the builders to only work on conversions

File: test/mlir/conversion/onnx_to_tosa/Math/Elementwise.mlir (hunk 9 above)
Comment (ttjost) [reply test/mlir/conversion/onnx_to_tosa/Math/Elementwise.mlir:255]: BTW, there was a bug. =)

File: src/Conversion/ONNXToTOSA/Math/Elementwise.cpp (hunk 2 above)
Comment (ttjost) [reply src/Conversion/ONNXToTOSA/Math/Elementwise.cpp:253]: Used `TosaBuilder::mul`

File: src/Conversion/ONNXToTOSA/Math/Elementwise.cpp (hunk 3 above)
Comment (ttjost) [reply src/Conversion/ONNXToTOSA/Math/Elementwise.cpp:268]: I think this can be closed for reasons I mentioned earlier.

**Review by TinaAMD — APPROVED** [review]
LGTM
